src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -32,34 +36,47 @@ def from_lockfile(cls, lockfile: bytes) -> LockfileMetadata:
             elif in_metadata_block:
                 metadata_lines.append(line[2:])
 
+        if not metadata_lines:
+            # TODO(#12314): Add a good error.
+            raise InvalidLockfileError("")
+
         try:
             metadata = json.loads(b"\n".join(metadata_lines))
         except json.decoder.JSONDecodeError:
             # TODO(#12314): Add a good error.
-            raise
-
-        T = TypeVar("T")
+            raise InvalidLockfileError("")
 
-        def coerce(t: Callable[[Any], T], key: str) -> T | None:
-            """Gets a value from `metadata`, coercing it to type `t` if not `None`."""
-            v = metadata.get(key, None)
+        def get(key: str) -> Any:
             try:
-                return t(v) if v is not None else None
-            except Exception:
-                # TODO: this should trigger error/warning behavior
```

Comment:
I think it should always be an error if the metadata is malformed, rather than a warning. That `[python-setup]` option imo is only whether to warn vs error if the metadata is _stale_, but not if the metadata is missing or malformed.